h2o-automl/src/main/java/ai/h2o/automl/AutoML.java

```diff
@@ -684,671 +493,39 @@ private int nextInstanceCounter(String algoName, NonBlockingHashMap<String, Inte
-   * @param parms the model builder params
-   * @param ignoreLimits (defaults to false) whether or not to ignore the max_models/max_runtime constraints
-   * @return a started training model
-   */
-  Job<Model> trainModel(Key<Model> key, WorkAllocations.Work work, Model.Parameters parms, boolean ignoreLimits) {
-    if (exceededSearchLimits(work, key == null ? null : key.toString(), ignoreLimits)) return null;
-
-    Algo algo = work.algo;
-    String algoName = ModelBuilder.algoName(algo.urlName());
-
-    if (null == key) key = modelKey(algoName);
-
-    Job<Model> job = new Job<>(key, ModelBuilder.javaName(algo.urlName()), algoName);
-    ModelBuilder builder = ModelBuilder.make(algo.urlName(), job, key);
-    Model.Parameters defaults = builder._parms;
-    builder._parms = parms;
-
-    setCommonModelBuilderParams(builder._parms);
-    setStoppingCriteria(builder._parms, defaults, true);
-
-    // override model's max_runtime_secs to ensure that the total max_runtime doesn't exceed expectations
-    if (ignoreLimits)
-      builder._parms._max_runtime_secs = 0;
-    else if (builder._parms._max_runtime_secs == 0)
-      builder._parms._max_runtime_secs = timeRemainingMs() / 1e3;
-    else
-      builder._parms._max_runtime_secs = Math.min(builder._parms._max_runtime_secs, timeRemainingMs() / 1e3);
-
-    builder.init(false);          // validate parameters
-
-    Log.debug("Training model: " + algoName + ", time remaining (ms): " + timeRemainingMs());
-    try {
-      return builder.trainModelOnH2ONode();
-    } catch (H2OIllegalArgumentException exception) {
-      eventLog().warn(Stage.ModelTraining, "Skipping training of model "+key+" due to exception: "+exception);
-      return null;
-    }
-  }
-
-  private Key<Grid> gridKey(String algoName) {
-    return Key.make(algoName + "_grid_" + nextInstanceCounter(algoName, gridInstanceCounters) + "_AutoML_" + timestampFormatForKeys.format(this.startTime));
+  Key<Grid> gridKey(String algoName, boolean with_counter) {
+    String counterStr = with_counter ? "_" + nextInstanceCounter(algoName, gridInstanceCounters) : "";
+    return Key.make(algoName + "_grid_" + counterStr + "_AutoML_" + timestampFormatForKeys.format(this.startTime));
   }
 
-  private void addGridKey(Key<Grid> gridKey) {
+  void addGridKey(Key<Grid> gridKey) {
     gridKeys = Arrays.copyOf(gridKeys, gridKeys.length + 1);
     gridKeys[gridKeys.length - 1] = gridKey;
   }
 
-  /**
-   * Do a random hyperparameter search.  Caller must eventually do a <i>get()</i>
-   * on the returned Job to ensure that it's complete.
-   * @param gridKey optional grid key
-   * @param work  the allocated work item: used to check various executions limits and to distribute remaining work (time + models)
-   * @param baseParms ModelBuilder parameter values that are common across all models in the search
-   * @param searchParms hyperparameter search space
-   * @return the started hyperparameter search job
-   */
```

Comment:
common `hyperparameterSearch` logic moved with barely no changes to `TrainingStep.GridStep` inner class.